tempto-core/src/main/java/com/teradata/tempto/s3/S3Client.java

```diff
@@ -0,0 +1,384 @@
+            // (indicates that the folder is located in this folder) then include in the output
+            if (slashCount == 0 || (slashCount == 1 && objectName.substring(objectName.length() - 1).equals("/"))) {
+                objects.add(object);
+            }
+        }
+
+        return objects;
+    }
+
+    /**
+     * Method to return if file or folder already exists
+     *
+     * @param path  path of file or folder
+     * @return      true/false
+     */
+    public boolean exists(String path)
+    {
+        return client.doesObjectExist(bucketName, path);
+    }
+
+    /**
+     * Method to create folder in S3
+     *
+     * @param path  folder path to create
+     */
+    public void createFolder(String path)
+    {
+        // If no forward slash at end of path, add one
+        if (!path.substring(path.length() - 1).equals("/")) {
+            path += "/";
+        }
+
+        // If folder already exists, throw exception
+        if (exists(path)) {
+            throw new RuntimeException("Folder already exists - could not create folder: " + path);
+        }
+
+        // Create meta-data for your folder and set content-length to 0
+        ObjectMetadata metadata = new ObjectMetadata();
+        metadata.setContentLength(0);
+
+        // Create empty content
+        InputStream emptyContent = new ByteArrayInputStream(new byte[0]);
+
+        // Create a PutObjectRequest passing the folder name suffixed by /
+        PutObjectRequest putObjectRequest = new PutObjectRequest(bucketName,
+                path, emptyContent, metadata);
+
+        // Send request to S3 to create folder
+        client.putObject(putObjectRequest);
+
+        logger.debug("Created S3 folder {}", path);
+    }
+
+    /**
+     * Method to delete folder in S3.  This method calls the main deleteFolder method, setting the recursive
+     * flag to false
+     *
+     * @param path  folder path to delete
+     */
+    public void deleteFolder(String path)
+    {
+        deleteFolder(path, false);
+    }
+
+    /**
+     * Method to delete folder.  If recursive flag is set to true, all objects underneath this folder
+     * will also be deleted
+     *
+     * @param path      folder path to delete
+     * @param recursive flag if set to true will delete all objects underneath
+     */
+    public void deleteFolder(String path, boolean recursive)
+    {
+        // If no forward slash at end of path, add one
+        if (!path.substring(path.length() - 1).equals("/")) {
+            path += "/";
+        }
+
+        // Get all objects in this folder
+        ArrayList<String> objects = getAllObjects(path);
+
+        if (!recursive && objects.size() > 0) {
+            throw new RuntimeException("Folder is not empty - could not remove folder: " + path);
+        } else if (objects.size() > 0) {
+            for(String objectName : objects) {
+                deleteObject(objectName);
```

Comment:
is the order of deletions here proper. I assume that `deleteObject` is not recursive, as otherwise we would not need to do this iteration here at all. So the question is are paths on `objects` list ordered in such way that we will delete children first and parent later?